// atatus-nodejs/lib/metrics/index.js
'use strict'

const MetricsRegistry = require('./registry')

const registrySymbol = Symbol('metrics-registry')
const agentSymbol = Symbol('metrics-agent')
const METRICSET_INTERVAL = 60

class NoopLogger {
  debug () { }
  error () { }
  fatal () { }
  info () { }
  trace () { }
  warn () { }
}

class Metrics {
  constructor (agent) {
    this[agentSymbol] = agent
    this[registrySymbol] = null
  }

  start () {
    // const metricsInterval = this[agentSymbol]._conf.metricsInterval
    const metricsInterval = METRICSET_INTERVAL
    this[registrySymbol] = new MetricsRegistry(this[agentSymbol], {
      reporterOptions: {
        defaultReportingIntervalInSeconds: metricsInterval,
        enabled: metricsInterval !== 0,
        unrefTimers: true,
        logger: new NoopLogger()
      }
    })
  }

  stop () {
    if (this[registrySymbol]) {
      this[registrySymbol].shutdown()
      this[registrySymbol] = null
    }
  }

  getOrCreateCounter (...args) {
    return this[registrySymbol].getOrCreateCounter(...args)
  }
}

module.exports = Metrics
